apps/designer/app/routes/$.tsx

```diff
@@ -0,0 +1,128 @@
+import { type LoaderFunction, redirect, MetaFunction } from "@remix-run/node";
+import { useLoaderData } from "@remix-run/react";
+import { InstanceRoot, Root } from "@webstudio-is/react-sdk";
+import { loadCanvasData, type CanvasData } from "~/shared/db";
+import config from "~/config";
+import { db } from "@webstudio-is/project/index.server";
+import env, { Env } from "~/env.server";
+import { sentryException } from "~/shared/sentry";
+import { Canvas } from "~/canvas";
+
+type Mode = "edit" | "preview" | "published";
+
+type Data =
+  | (CanvasData & { env: Env; mode: Mode })
+  | { errors: string; env: Env };
+
+type ProjectIdentifier = { type: "id" | "domain"; value: string };
+
+export const isCanvasRequest = (
+  request: Request
+): false | { projectId: ProjectIdentifier; mode: Mode; pathname: string } => {
+  const url = new URL(request.url);
+
+  const pathname = url.pathname;
+
+  if (Object.values(config).some((path) => pathname.startsWith(path))) {
+    return false;
+  }
+
+  let projectId: ProjectIdentifier | undefined = undefined;
+
+  // @todo all this subdomain logic is very hacky
+  const host =
+    request.headers.get("x-forwarded-host") ||
+    request.headers.get("host") ||
+    "";
+  const [userDomain, wstdDomain] = host.split(".");
+  if (wstdDomain === "wstd" || wstdDomain?.includes("localhost")) {
+    projectId = { type: "domain", value: userDomain };
+  }
+
+  if (projectId === undefined) {
+    const projectIdParam = url.searchParams.get("projectId");
+    if (projectIdParam !== null) {
+      projectId = { type: "id", value: projectIdParam };
+    }
+  }
+
+  if (projectId === undefined) {
+    return false;
+  }
+
+  const modeParam = url.searchParams.get("mode");
```

Comment:
Maybe a bit more organized

if (modes.includes(mode) === false) { 
   throw new Error
}
return {projectId, mode, pathname}

Reply:
this would handle the case if mode param is wrong, so we don't fallback to published